irohad/consensus/yac/impl/yac.cpp

```diff
@@ -139,13 +143,39 @@ namespace iroha {
       void Yac::onState(std::vector<VoteMessage> state) {
         std::unique_lock<std::mutex> guard(mutex_);
 
-        removeUnknownPeersVotes(state);
+        removeUnknownPeersVotes(state, getCurrentOrder());
         if (state.empty()) {
           log_->debug("No votes left in the message.");
           return;
         }
 
         if (crypto_->verify(state)) {
+          auto &proposal_round = getRound(state);
+
+          if (proposal_round.block_round > round_.block_round) {
+            guard.unlock();
+            log_->info("Pass state from future for {} to pipeline",
+                       proposal_round);
+            notifier_.get_subscriber().on_next(FutureMessage{std::move(state)});
+            return;
+          }
+
+          if (proposal_round.block_round < round_.block_round) {
+            log_->info("Received state from past for {}, try to propagate back",
+                       proposal_round);
+            tryPropagateBack(state);
+            guard.unlock();
+            return;
+          }
+
+          // filter votes with peers from cluster order to avoid the case when
+          // alternative peer is not present in cluster order
+          removeUnknownPeersVotes(state, cluster_order_);
```

Comment:
this call is needed only once and should be avoided other times. I would make
```suggestion
        if (alternative_order_) {
          removeUnknownPeersVotes(state, cluster_order_);
        }
```